ugc/loader.cpp

```diff
@@ -21,17 +21,14 @@ UGC Loader::GetUGC(FeatureID const & featureId)
   if (!value.m_cont.IsExist(UGC_FILE_TAG))
     return {};
 
-  if (m_currentMwmId != featureId.m_mwmId)
+  UGC ugc;
+  binary::UGCDeserializer deserializer;
   {
-    m_currentMwmId = featureId.m_mwmId;
-    m_d = binary::UGCDeserializer();
+    std::lock_guard<std::mutex> lock(m_mutex);
+    auto deserializer = m_deserializers[featureId.m_mwmId];
   }
-
   auto readerPtr = value.m_cont.GetReader(UGC_FILE_TAG);
-
-  UGC ugc;
-  if (!m_d.Deserialize(*readerPtr.GetPtr(), featureId.m_index, ugc))
-    return {};
+  deserializer.Deserialize(*readerPtr.GetPtr(), featureId.m_index, ugc);
```

Comment:
This won't work, because deserializer has an internal structure, and Deserialize method is not thread safe? I suggest you to do following:
```c++
struct Entry
{
  binary::UGCDeserializer m_deserializer;
  std::mutex m_guard;
};

map<MwmSet::MwmId, shared_ptr<Entry>> m_deserializers;
```
Then:
```c++
shared_ptr<Entry> entry;
{
  lock_guard<mutex> lock(m_mutex);
  entry = m_deserializers[featureId.m_mwmId];
}
if (entry)
{
  lock_guard<mutex> guard(entry.m_mutex);
  ... deserialize all what you need here ....
}
```

Reply:
Да, в моем варианте кэширование данных будет производиться внутри локальных копий десериализаторов. Переделал по предложенному варианту